src/Umbraco.Web.BackOffice/Middleware/BasicAuthAuthenticationMiddleware.cs

```diff
@@ -0,0 +1,101 @@
+using System;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.Extensions.DependencyInjection;
+using Umbraco.Cms.Core;
+using Umbraco.Cms.Core.Services;
+using Umbraco.Cms.Web.BackOffice.Security;
+using Umbraco.Extensions;
+
+namespace Umbraco.Cms.Web.Common.Middleware
+{
+    /// <summary>
+    ///     Ensures that preview pages (front-end routed) are authenticated with the back office identity appended to the
+    ///     principal alongside any default authentication that takes place
+    /// </summary>
+    public class BasicAuthAuthenticationMiddleware : IMiddleware
+    {
+        private readonly IRuntimeState _runtimeState;
+        private readonly IBasicAuthService _basicAuthService;
+
+        public BasicAuthAuthenticationMiddleware(
+            IRuntimeState runtimeState,
+            IBasicAuthService basicAuthService)
+        {
+            _runtimeState = runtimeState;
+            _basicAuthService = basicAuthService;
+        }
+
+        /// <inheritdoc />
+        public async Task InvokeAsync(HttpContext context, RequestDelegate next)
+        {
+            if (_runtimeState.Level < RuntimeLevel.Run || !_basicAuthService.IsBasicAuthEnabled())
+            {
+                await next(context);
+                return;
+            }
+
+            var clientIPAddress = context.Connection.RemoteIpAddress;
+            if (_basicAuthService.IsIpAllowListed(clientIPAddress))
+            {
+                await next(context);
+                return;
+            }
+
+            AuthenticateResult authenticateResult = await context.AuthenticateBackOfficeAsync();
+            if (authenticateResult.Succeeded)
+            {
+                await next(context);
+                return;
+            }
+
+
+            string authHeader = context.Request.Headers["Authorization"];
+            if (authHeader != null && authHeader.StartsWith("Basic"))
+            {
+                //Extract credentials
+                var encodedUsernamePassword = authHeader.Substring(6).Trim();
```

Comment:
Could move this out into a helper method, or perhaps an extension method on `HttpRequest` called `.TryGetBasicAuthCredentials(out string username, out string password)`, and then can unit test.